### PR TITLE
GCW 2216- Stock Dashboard orders count

### PR DESCRIPTION
**Pull request by abulsayyad123**
Hi Team, 
This is PR for jira ticket GCW-2216 - Stock Dashboard orders count and latest orders. https://jira.crossroads.org.hk/browse/GCW-2216. 
This PR includes:
* Orders count of Priority and Non-priority type.
* Specs for the scenarios.
Please review the PR and let me know if any change is required.
Thanks

### CODE REVIEW

File: app/models/order.rb [hunk 1]
```diff
@@ -297,15 +309,11 @@ def self.join_order_associations
...
-    active_orders
-    .select("DISTINCT ON (orders.id) orders.id AS key,  versions.created_at AS recently_used_at").
-    joins("INNER JOIN versions ON ((object_changes -> 'order_id' ->> 1) = CAST(orders.id AS TEXT))").
-    joins("INNER JOIN packages ON (packages.id = versions.item_id AND versions.item_type = 'Package')").
-    where(" versions.event = 'update' AND
-      (object_changes ->> 'order_id') IS NOT NULL AND
-      CAST(whodunnit AS integer) = ? AND
-      versions.created_at >= ? ", user_id, 15.days.ago).
-    order("key, recently_used_at DESC")
+    active_orders.goodcity_orders.joins(:versions).where(versions: {id: Version.select(:id).where(whodunnit: user_id.to_s, item_type: "Order")}).distinct.order(updated_at: :desc).limit(5)
```
Comment (steveyken) [inline app/models/order.rb]: this line is too long and hard to read... can you shorten. also, what is the performance of this compare to the previous code?

File: app/models/order.rb [hunk 2]
```diff
@@ -69,6 +69,17 @@ class Order < ActiveRecord::Base
 
   scope :goodcity_orders, -> { where(detail_type: 'GoodCity') }
 
+  def filter(opts = {})
+    query = FilterQueryBuilder.new
```
Comment (steveyken) [inline app/models/order.rb]: @patrixr has moved the FilterQueryBuilder to another class file - can you rebase and update here?

File: app/controllers/api/v1/orders_controller.rb [hunk 3]
```diff
@@ -83,7 +83,7 @@ def update
 
       def recent_designations
         records = Order.recently_used(User.current_user.id)
-        render json: order_response(records)
+        render json: order_response(records.to_a)
```
Comment (steveyken) [inline app/controllers/api/v1/orders_controller.rb]: do we really need `to_a` here?

File: app/controllers/api/v1/orders_controller.rb [hunk 4]
```diff
@@ -96,8 +96,23 @@ def destroy
         render json: {}
       end
 
+      def summary
+        active_orders = Order.orders_group_by_state(false)
```
Comment (steveyken) [inline app/controllers/api/v1/orders_controller.rb]: I'm not a huge fan of passing boolean arguments as it's not clear what they relate to.

Could something like `Order.orders_group_by_state(priority: false)` work better? Makes it clearer what the true and false are relating too.

File: app/models/order.rb (hunk 1 above)
Comment (abulsayyad123) [reply app/models/order.rb]: Previous query was not giving expected result and it even was not working(I suspect). I tried to add change in that query itself but it was difficult to debug. That is why I rewrite the query again. I'll look into improve and shorten the query.

**Comment by steveyken** [comment]
Ok, best to write tests to prove the failure and then that your solution
fixes it. TDD ftw!

**Comment by abulsayyad123** [comment]
Recent designation will be another PR. As it has some edge case scenarios so will be available in latter PR along with all specs covered.

File: app/controllers/api/v1/orders_controller.rb [hunk 5]
```diff
@@ -96,8 +96,23 @@ def destroy
         render json: {}
       end
 
+      def summary
+        active_orders = Order.orders_group_by_state(is_priority: false)
+        active_orders_count = orders_count(active_orders)
+
+        priority_orders = Order.orders_group_by_state(is_priority: true).transform_keys{ |key| "priority_".concat(key) }
```
[the remark this answers is not in the document]
Comment (namrataukirde) [reply app/controllers/api/v1/orders_controller.rb]: @abulsayyad123 it will be good to keep these 2 methods different. It's difficult to understand. 

e.g first method `active_orders_count` and other is `priority_active_orders_count ` . Then summary method will be like: 

`active_orders_count.merge!(priority_active_orders_count)
        render json: active_orders_count`

File: app/models/concerns/order_filtering.rb [hunk 6]
```diff
@@ -3,7 +3,7 @@ module OrderFiltering
   extend ActiveSupport::Concern
 
   def is_priority?
-    self.class.priority.exists?(self)
+    self.class.priority.exists?(self.id)
```
Comment (namrataukirde) [inline app/models/concerns/order_filtering.rb]: we can remove `self` here.

File: app/models/order.rb [hunk 7]
```diff
@@ -308,6 +308,10 @@ def self.recently_used(user_id)
     order("key, recently_used_at DESC")
   end
 
+  def self.orders_group_by_state(is_priority: false)
```
Comment (namrataukirde) [inline app/models/order.rb]: It will be good to add specs for this.

File: app/controllers/api/v1/orders_controller.rb [hunk 8]
```diff
@@ -96,8 +96,23 @@ def destroy
         render json: {}
       end
 
+      def summary
```
Comment (namrataukirde) [inline app/controllers/api/v1/orders_controller.rb]: Do we already have specs for this endpoint ?

Comment (abulsayyad123) [reply app/controllers/api/v1/orders_controller.rb]: Yes, I have added specs for this endpoint.

File: app/models/order.rb (hunk 7 above)
Comment (abulsayyad123) [reply app/models/order.rb]: Added.
